Port gen_opt_builtin* and getblockparamproxy

noahgibbs opened the pull request:
...

noahgibbs committed:
Port gen_opt_invokebuiltin functions

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -4996,46 +5000,44 @@ fn gen_invokebuiltin(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock,
 
     // ec, self, and arguments
-    if (bf->argc + 2 > NUM_C_ARG_REGS) {
+    if bf_argc + 2 > (C_ARG_REGS.len().try_into().unwrap()) {
```
I feel like it's very verbose to use `try_into().unwrap()` everywhere. In most cases we know the value can't overflow so it would be fine to use `as`. In this specific case the array length is a constant.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -4972,19 +4975,20 @@ fn gen_invokebuiltin(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock,
 
     // Push the return value
-    ctx.stack_pop(bf->argc);
+    ctx.stack_pop(bf_argc.try_into().unwrap());
```
We could use `as` here too because we've already checked that `bf_argc` is less than a small constant.

maximecb reviewed (commented):
Looks good with minor comments.

noahgibbs commented:
Okay. Think I've addressed those.

maximecb reviewed (approved):
We're making a lot of progress, in large part thanks to you :)

XrXr committed:
Slightly better numeric casting

Casting to usize once at the top removes one cast. We also did not check
that argc is non negative before this change. I guess the C code is
using `int` for the field even though it's never negative.

Feel free to revert / make changes if you disagree that this is an
improvement.

noahgibbs commented:
No trouble :-)
